src/Database/PostgreSQL/TransactionalStore.hs

```diff
@@ -21,74 +22,97 @@ module Database.PostgreSQL.TransactionalStore
 import           Control.Applicative
 #endif
 import           Control.Monad.Reader
+import           Control.Monad.Trans.Control
 import           Data.Int
-import qualified Database.PostgreSQL.Simple         as Postgres
-import qualified Database.PostgreSQL.Simple.Transaction as Postgres.Transaction
-import qualified Database.PostgreSQL.Simple.Types   as PGTypes
+import qualified Database.PostgreSQL.Simple             as Postgres
 import           Database.PostgreSQL.Simple.FromRow
 import           Database.PostgreSQL.Simple.ToRow
+import qualified Database.PostgreSQL.Simple.Transaction as Postgres.Transaction
+import qualified Database.PostgreSQL.Simple.Types       as PGTypes
 
-newtype PGTransaction a =
-    PGTransaction (ReaderT Postgres.Connection IO a)
+newtype PGTransactionT m a =
+    PGTransactionT (ReaderT Postgres.Connection m a)
     deriving ( Functor
              , Applicative
              , Monad
-             , MonadIO
+             , MonadTrans
              , MonadReader Postgres.Connection
+             , MonadIO
              )
 
-runPGTransaction' :: MonadIO m
+type PGTransaction = PGTransactionT IO
+
+runPGTransaction' :: MonadBaseControl IO m
```

Comment:
`withTransactionLevel` takes an `IO a` as its third argument, the query action to be run within a transaction. `MonadBaseControl IO` allows us to get an `IO a` value from `ReaderT` and "wrap" the result of `withTransactionLevel` back up in `ReaderT`. `MonadIO` only allows you to naturally transform an `IO` functor, it doesn't allow you to extract an `IO a` from it.

 I definitely recommend http://www.yesodweb.com/book/monad-control as a decent explanation of the movitation and workings of `MonadBaseControl`